_posts/2023-08-25-migrer-de-commonjs-vers-esm.md

```diff
@@ -0,0 +1,421 @@
+---
+layout: post
+title:  "Migrer de CommonJS vers ECMAScript Modules #1"
+date: 2023-08-25 09:49:00 +0200
+categories: javascript
+excerpt: "Récit d'une aventure : migrer 400kLoc NodeJs en deux mois - #1: les modules"
+cover:
+  image: "cjs-to-esm.png"
+  source_url: https://raw.githubusercontent.com/wessberg/cjstoesm/master/documentation/asset/logo.png
+authors: pierre_top
+---
+
+## Y'a quoi au menu ?
+Cet article partage avec vous deux mois d'efforts (et de doutes, aussi) pour migrer une base de code NodeJS de 400 kLoC. Spoiler : nous y sommes arrivés !
+
+Vous y trouverez de l'histoire, de la syntaxe, des considérations techniques et organisationnelles, de l'outillage et du monitoring : il y en a pour tous les goûts.
+
+Voilà, en entrée, le résumé du challenge à accomplir :
+- [comprendre](#une-histoire-de-modules) ce qu'est un système de module ;
+- se souvenir [de ce qu'il en est](#les-modules-côté-javascript) en Javascript ;
+- savoir si ESM [peut t'être utile](#pourquoi-migrer-en-esm-) ;
+- comprendre [la syntaxe ESM](#un-peu-de-syntaxe) ;
+- comment [modifier le code](#modifier-le-code) (la [syntaxe](#un-peu-de-syntaxe) est un pré-requis).
+
+Si vous connaissez déjà ESM et les AST, attendez plutôt la sortie de l'article [en plat principal](https://github.com/1024pix/pix-engineering/pull/42), à savoir comment nous avons mené la migration.
+
+## Une histoire de modules
+
+Dans la plupart des langages de programmation (C, Java, JavaScript...), le fichier de code (source) est un élément important. Il est implémenté dans le système de fichiers de l'OS, par un fichier texte. Il y a quelque chose de fondamentalement réconfortant à cela. Utiliser un langage qui stocke le code source dans des fichiers aux noms cryptiques et au contenu binaire, c'est une perte d'autonomie.
+
+Pourquoi ? Parce que ranger des fichiers dans des dossiers est une métaphore puissante : elle permet de tracer des frontières dans sa tête, et d'oublier les particularités pour un moment. C'est la même chose qui se passe au niveau en-dessous lorsqu'on crée des fonctions.
+
+> Aside from the computer itself, the routine is the single greatest invention in computer science. (...) Create a routine to hide information so that you won't need to think about it. (...) Without the abstractive power of routines, complex programs would be impossible to manage intellectually.
```

Comment:
Juste un HS en passant `We don’t write things down to remember them. We write them down to forget.`
https://www.reproof.app/blog/notes-apps-help-us-forget

Reply:
Cela s'applique très bien aux journaux personnels maintenant que tu le dis